src/server/supplements/har/harTracer.ts

```diff
@@ -152,38 +155,50 @@ export class HarTracer {
         receive: -1
       },
     };
-    if (request.redirectedFrom()) {
-      const fromEntry = this._entries.get(request.redirectedFrom()!)!;
-      fromEntry.response.redirectURL = request.url();
-    }
+
     this._log.entries.push(harEntry);
     this._entries.set(request, harEntry);
   }
 
-  private _onResponse(page: Page, response: network.Response) {
+  private async _onRequestFinished(page: Page, request: network.Request) {
     const pageEntry = this._pageEntries.get(page)!;
-    const harEntry = this._entries.get(response.request())!;
+    const harEntry = this._entries.get(request)!;
     // Rewrite provisional headers with actual
-    const request = response.request();
     harEntry.request.headers = request.headers().map(header => ({ name: header.name, value: header.value }));
     harEntry.request.cookies = cookiesForHar(request.headerValue('cookie'), ';');
-    harEntry.request.postData = postDataForHar(request) || undefined;
+
+    const response = await request.response();
+
+    if (!response) return;
+
+    const httpVersion = response.protocol() ? response.protocol() : FALLBACK_HTTP_VERSION;
+    const transferSize = response.encodedDataLength();
+    const headersSize = calcResponseHeadersSize(httpVersion, response.status(), response.statusText(), response.headers());
```

Comment:
Note that in Chromium we have `headersText` arriving in `Network.responseReceivedExtraInfo`, so that you don't have to calculate it. Let's add a TODO to use that.

Reply:
@dgozman I think a TODO would be fine, right now my solution is useful as it is uniform for all supported browsers.